File: demos/02-havoc-coverage/concrete_type_safe/add_one_disproved.rs
/*
DEMO: Disproved — concrete struct method that mutates a `static mut` global.

Mirrors demos/02-havoc-coverage/concrete_type_safe/add_one_disproved.cpp.
The original C++ uses a `SusLog` whose log() method clobbers
`super_important`. Here we do the same thing in Rust: a concrete
struct's method writes -1 to a static mut global, which triggers the
bail-out branch in add_one and returns 12 instead of x + 1.
*/

static mut SUPER_IMPORTANT: i32 = 7;

struct SusLog;

impl SusLog {
    fn log(&self, _msg: &str) {
        // SAFETY: single-threaded demo; demonstrates mutating a global.
        unsafe { SUPER_IMPORTANT = -1; }
    }
}

fn add_one(x: u32) -> u32 {
    let logger = SusLog;
    logger.log("Adding one to x");

    // SAFETY: same as above.
    if unsafe { SUPER_IMPORTANT } == -1 {
        return 12;
    }
    x + 1
}

File: tests/e2e/cases/02-havoc-coverage/concrete_type_safe/add_one_disproved.cpp
#include <cstdint>
#include <cstdio>
#include <cstdlib>

int super_important = 7;

/*
DEMO: Disproved because the concrete type's method clobbers global memory

Unlike the disproved version which uses OkLog (safe implementation),
this version uses SusLog directly. Even though we're calling on
a concrete type (no vtable havoc), SAW verifies the ACTUAL
implementation — and SusLog::log() modifies super_important.

The proof fails because the branch `super_important == -1` is
reachable and returns 12 instead of x + 1.

COMPARE: add_one_verified.cpp uses OkLog — same non-const method,
but the concrete implementation doesn't touch global state.
*/

class ILog {
public:
    virtual void log(const char* message) = 0;
};

class OkLog : public ILog {
public:
    void log(const char* message) override {
        printf("OK: %s\n", message);
    }
};

class SusLog : public ILog {
public:
    void log(const char* message) override {
        super_important = -1;  // clobbers global!
        printf("SUS: %s\n", message);
    }
};

// Spec: add_one(x) = x + 1
uint32_t add_one(uint32_t x) {
    // Concrete type — SAW verifies SusLog::log() directly,
    // and finds it DOES clobber super_important
    SusLog logger;

    logger.log("Adding one to x");

    // SAW proves this branch IS reachable — returns 12, not x+1
    if (super_important == -1) {
        return 12;
    }

    return x + 1;
}
